### PR TITLE
simplify copy attr  speed up metatensor

### PR DESCRIPTION
**Pull request by wyli**
### Description
remove the recursive deepcopy to speed up operations

this PR:
Screenshot 2022-07-16 at 13 27 09

dev branch:
Screenshot 2022-07-16 at 06 57 36

0.9.0:
Screenshot 2022-07-16 at 07 08 34

### Status
**Ready/Work in progress/Hold**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [x] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

File: monai/data/utils.py [hunk 1]
```diff
@@ -588,11 +588,11 @@ def decollate_batch(batch, detach: bool = True, pad=True, fill_value=None):
         if isinstance(batch, MetaObj):
             for t, m in zip(out_list, decollate_batch(batch.meta)):
                 if isinstance(t, MetaObj):
-                    t.meta = m
+                    t.meta = deepcopy(m)
```
Comment (Nic-Ma) [inline monai/data/utils.py:591]: May I know why add `deepcopy` here? I think the decollated items of `batch.meta` should be dispatched to every item in `out_list`? Or Maybe I misundertand this logic here.

Thanks.

File: monai/data/meta_obj.py [hunk 2]
```diff
@@ -103,71 +103,30 @@ def flatten_meta_objs(*args: Iterable):
...
         """
-        Copy metadata from an iterable of `MetaObj` instances. For a given attribute, we copy the
-        adjunct data from the first element in the list containing that attribute.
+        Copy metadata from a `MetaObj` or an iterable of `MetaObj` instances.
 
         Args:
             input_objs: list of `MetaObj` to copy data from.
-
+            copy_attr: whether to copy each attribute with `MetaObj.copy_item`.
+                note that if the attribute is a nested list or dict, only a shallow copy will be done.
```
Comment (Nic-Ma) [inline monai/data/meta_obj.py:122]: Seems even if `copy_attr = True`, it still shallow copy the numpy array or tensor?
So if I really want to deepcopy, can I use `a = MetaTensor(...), MetaTensor b = deepcopy(a)` or `b.meta = deepcopy(a.meta)`?

Thanks.

Comment (wyli) [reply monai/data/meta_obj.py:122]: that's a good point, now it's updated to deepcopy in two cases when a is a MetaTensor, `b = MetaTensor(a)` and `b = a.clone()`, what do you think?

File: monai/data/utils.py (hunk 1 above)
Comment (wyli) [reply monai/data/utils.py:591]: this is to ensure all the samples in a batch don't share the meta info (to be consistent with applied_operations). I think it's robust, and doesn't require much cpu time as it's done only once per tensor.

File: monai/data/meta_obj.py (hunk 2 above)
Comment (Nic-Ma) [reply monai/data/meta_obj.py:122]: OK, the solution looks good to me. Please document it clearly, I feel it's easy to be confused..

Thanks.

File: monai/data/utils.py (hunk 1 above)
Comment (wyli) [reply monai/data/utils.py]: thanks -- that's removed in the latest commit

**Review by Nic-Ma — APPROVED** [review]

**Comment by rijobro** [comment]
Hi, it seems like the difference in your screenshot is around 0.1s  between slowest and fastest (1.93s vs 2.04s). Are we sure this isn't random fluctuations?

**Comment by wyli** [comment]
Yes @rijobro, I run `python tests/test_integration_fast_train.py` the total time is 30s vs 20s with my workstations, it's also the case for the relevant demos: https://github.com/Project-MONAI/tutorials/pull/786#issuecomment-1186073996.

**Pull request merged**